compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -1793,6 +1799,26 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
             }
         }
 
+        // if next_seg is present, let's filter everything that does not continue the path
+        if let Some(following_seg) = following_seg {
+            names.retain(|suggestion| match suggestion.res {
+                Res::Def(DefKind::Struct | DefKind::Enum | DefKind::Union, _) => {
+                    // this is not totally accurate, but mostly works
```

Comment:
Let's mark this with `FIXME`, to highlight that we might want to revisit the heuristic.